File: prime_factors.py
from math import gcd

def solve_for(n):
    (p, q)  = (1, n)
    coprime = 1
    while p >= n or q >= n:
        coprime = next_coprime_of(n, coprime)
        period  = even_period_of(n, coprime)
        (p, q)  = factors_of(n, coprime, period)
    return (p, q)

def next_coprime_of(n, coprime):
    while True:
        coprime += 1
        if gcd(n, coprime) == 1: break
    return coprime

def even_period_of(n, base):
    period = 2
    while (base ** period) % n != 1:
        period += 2
    return period

def factors_of(n, base, period):
    base_to_half_period = base ** (period // 2)
    p = gcd(n, base_to_half_period + 1)
    q = gcd(n, base_to_half_period - 1)
    return (p, q)

for n in [15, 35, 65, 91]:
    print("The factors of {} are {}".format(n, solve_for(n)))

File: PrimeFactors.hs
factorize :: Integer -> (Integer, Integer)
factorize n =
  let coprimes = [ m | m <- [2..n-1], gcd m n == 1 ]
      periods  = map (evenPeriodOf n) coprimes
      factors  = filter validFactors $ zipWith (factorsOf n) coprimes periods
  in head factors
  where
    validFactors (p, q) = p < n && q < n
   
evenPeriodOf :: Integer -> Integer -> Integer
evenPeriodOf n base = go 2
  where
    go r | (base ^ r) `mod` n == 1 = r
         | otherwise = go (r+2)

factorsOf :: Integer -> Integer -> Integer -> (Integer, Integer)
factorsOf n base period = (p, q)
  where
    p = gcd (w + 1) n
    q = gcd (w - 1) n
    w = base ^ (period `div` 2)


main :: IO ()
main = mapM_ (print . factorize) [15, 35, 65, 91]